File: InsertionSort/Swift/Sources/InsertionSort/InsertionSort.swift
//
//  InsertionSort.swift
//  Created by Ardalan Samimi on 2018-05-29.
//
//  Insertion sort is an in-place sorting algorithm, which compares
//  each element in sequence and puts it in its proper place in the
//  sorted list.
//
//  Worst-case complexity: O(n^2), where n is the number of elements
//

public func insertionSort<T: Comparable>(_ array: [T], compareUsing compare: (T, T) -> Bool = (<)) -> [T] {
	var array = array
	let count = array.count

	for curIndex in 1..<count {
		var subIndex = curIndex
		let current  = array[curIndex]

		while subIndex > 0 && compare(current, array[subIndex - 1]) {
			array[subIndex] = array[subIndex - 1]
			subIndex -= 1
		}

		array[subIndex] = current
	}

	return array
}

File: InsertionSort/Java/Sources/InsertionSort/InsertionSort.java
//
//  InsertionSort.java
//  InsertionSort
//
//  Created by Ardalan Samimi on 2019-03-29.
//
package InsertionSort;

public class InsertionSort {

	public static void sort(int[] array) {
		int size = array.length;

		for (int i = 1; i < size; i++) {
			int j = i - 1;
			int key = array[i];

			while (j >= 0 && array[j] > key) {
				array[j + 1] = array[j];
				j -= 1;
			}

			array[j + 1] = key;
		}
	}

}
